local-volume/provisioner/pkg/deleter/deleter.go

```diff
@@ -46,60 +52,191 @@ func (d *Deleter) DeletePVs() {
 	case common.VolumeTypeBlock:
-		return fmt.Errorf("Not yet implemented")
+		return d.cleanupBlockPV(pv, mountPath, config)
 	default:
 		return fmt.Errorf("Unexpected volume type %q for deleting path %q", volType, pv.Spec.Local.Path)
 	}
 }
 
-func (d *Deleter) cleanupFileVolume(pv *v1.PersistentVolume, config common.MountConfig) error {
-	specPath := pv.Spec.Local.Path
-	relativePath, err := filepath.Rel(config.HostDir, specPath)
+func (d *Deleter) deleteFilePV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error {
+	err := d.ProcTable.MarkRunning(pv.Name)
+	if err != nil {
+		return err
+	}
+	go d.asyncDeleteFilePV(pv, mountPath, config)
+	return nil
+}
+
+func (d *Deleter) asyncDeleteFilePV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error {
+
+	defer d.ProcTable.MarkDone(pv.Name)
+	glog.Infof("Deleting PV file volume %q contents at hostpath %q, mountpath %q", pv.Name, pv.Spec.Local.Path,
+		mountPath)
+	if err := d.VolUtil.DeleteContents(mountPath); err != nil {
+		return err
+	}
+
+	// Remove API object
+	if err := d.APIUtil.DeletePV(pv.Name); err != nil {
+		// TODO: Does delete return an error if object has already been deleted?
+		deletingLocalPVErr := fmt.Errorf("Error deleting PV %q: %v", pv.Name, err.Error())
+		d.RuntimeConfig.Recorder.Eventf(pv, v1.EventTypeWarning, common.EventVolumeFailedDelete,
+			deletingLocalPVErr.Error())
+		glog.Error(deletingLocalPVErr)
+		return err
+	}
+
+	glog.Infof("Deleted PV %q", pv.Name)
+
+	return nil
+}
+
+// cleanupBlockPV is meant to execute cleanup concurrently with other such cleanup processes
+// cmdio is a param that can suppoly io to the script being invoked. Its main purpose currently is
+// to enable better testing of the code path with output validation etc.
+func (d *Deleter) cleanupBlockPV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error {
+	// Make absolutely sure here that we are not deleting anything outside of mounted dir
+	if !strings.HasPrefix(mountPath, config.MountDir) {
+		err := fmt.Errorf("Unexpected error pv %q mountPath %s but mount dir is %s", pv.Name, mountPath,
+			config.MountDir)
+		glog.Error(err)
+		return err
+	}
+
+	if len(config.BlockCleanerCommand) < 1 {
+		err := fmt.Errorf("Blockcleaner command was empty for pv %q ountPath %s but mount dir is %s", pv.Name,
+			mountPath, config.MountDir)
+		glog.Error(err)
+		return err
+	}
+
+	cleaningInfo := fmt.Errorf("Starting cleanup of Block PV %q, this may take a while", pv.Name)
+	d.RuntimeConfig.Recorder.Eventf(pv, v1.EventTypeNormal, common.VolumeDelete, cleaningInfo.Error())
+
+	err := d.ProcTable.MarkRunning(pv.Name)
+	if err != nil {
+		return err
+	}
+	go d.asyncCleanupBlockPV(pv, mountPath, config.BlockCleanerCommand[0], config.BlockCleanerCommand[1:]...)
+
+	return nil
+}
+
+func (d *Deleter) asyncCleanupBlockPV(pv *v1.PersistentVolume, blkdevPath string, exe string, exeArgs ...string) error {
```

Comment:
Can we have a common async cleanup method for both file and block PV?  It can take in a func() as an arg, that will call the file-specific methods or the block-specific methods.  Then we can have a common function that marks marks proc as running, mark as done, and delete PV object.